source/extensions/filters/network/kafka/serialization.h

```diff
@@ -0,0 +1,776 @@
+    const size_t data_consumed = std::min<size_t>(required_, remaining);
+    const size_t written = data_buf_.size() - required_;
+    memcpy(data_buf_.data() + written, buffer, data_consumed);
+    required_ -= data_consumed;
+
+    buffer += data_consumed;
+    remaining -= data_consumed;
+
+    if (required_ == 0) {
+      ready_ = true;
+    }
+
+    return length_consumed + data_consumed;
+  }
+
+  bool ready() const override { return ready_; }
+
+  NullableString get() const override {
+    return required_ >= 0 ? absl::make_optional(std::string(data_buf_.begin(), data_buf_.end()))
+                          : absl::nullopt;
+  }
+
+private:
+  constexpr static int16_t NULL_STRING_LENGTH{-1};
+
+  Int16Deserializer length_buf_;
+  bool length_consumed_{false};
+
+  int16_t required_;
+  std::vector<char> data_buf_;
+
+  bool ready_{false};
+};
+
+/**
+ * Deserializer of bytes value
+ * First reads length (INT32) and then allocates the buffer of given length
+ *
+ * From documentation:
+ * First the length N is given as an INT32. Then N bytes follow.
+ */
+class BytesDeserializer : public Deserializer<Bytes> {
+public:
+  /**
+   * Can throw EnvoyException if given bytes length is not valid
+   */
+  size_t feed(const char*& buffer, uint64_t& remaining) override {
+    const size_t length_consumed = length_buf_.feed(buffer, remaining);
+    if (!length_buf_.ready()) {
+      // break early: we still need to fill in length buffer
+      return length_consumed;
+    }
+
+    if (!length_consumed_) {
+      required_ = length_buf_.get();
+      if (required_ >= 0) {
+        data_buf_ = std::vector<unsigned char>(required_);
+      } else {
+        throw EnvoyException(fmt::format("invalid BYTES length: {}", required_));
+      }
+      length_consumed_ = true;
+    }
+
+    const size_t data_consumed = std::min<size_t>(required_, remaining);
+    const size_t written = data_buf_.size() - required_;
+    memcpy(data_buf_.data() + written, buffer, data_consumed);
+    required_ -= data_consumed;
+
+    buffer += data_consumed;
+    remaining -= data_consumed;
+
+    if (required_ == 0) {
+      ready_ = true;
+    }
+
+    return length_consumed + data_consumed;
+  }
+
+  bool ready() const override { return ready_; }
+
+  Bytes get() const override { return data_buf_; }
+
+private:
+  Int32Deserializer length_buf_;
+  bool length_consumed_{false};
+  int32_t required_;
+
+  std::vector<unsigned char> data_buf_;
+  bool ready_{false};
+};
+
+/**
+ * Deserializer of nullable bytes value
+ * First reads length (INT32) and then allocates the buffer of given length
+ * If length was -1, buffer allocation is omitted and deserializer is immediately ready (returning
+ * null value)
+ *
+ * From documentation:
+ * For non-null values, first the length N is given as an INT32. Then N bytes follow.
+ * A null value is encoded with length of -1 and there are no following bytes.
+ */
+class NullableBytesDeserializer : public Deserializer<NullableBytes> {
```

Comment:
honestly, `NullableBytesDeserializer` & `BytesDeserializer` (or `N~StringD~`, or `N~ArrayD~`) are so similar that they could inherit from a single class with some minor configuration object (strategy to tell us what to do with given `int length`) and output converter (if we store result as `optional<vector<T>` then BytesDeserializer needs to invoke unpack the option, while the NullableB~D~ does not
I'll wait if I've caused any worse errors first, then happy to change it if requested

Reply:
Up to you. In general this all looks great, so however you want to wind up, though I would personally favor less duplicated code if possible.